### PR TITLE
[Feature] configs: new config infra with pydantic, with config via env and dynamic rootfs support

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Description

> [!Note]
> With pydantic being introduced into the code base, the test CI is broken caused by coverage measuring setup, #263 is required for making the pytest with coverage work again. 

>[!Note]
> This PR implements the config infra described in [otaclient configs infra](https://tier4.atlassian.net/l/cp/e8q02QeD).

This PR introduces new features based on refactoring otaclient's config infrastructure with `pydantic` as follow.

### standalone `otaclient.configs` package for holding `otaclient` package scope configs

Configs for `otaclient` top-level modules(i.e., `app`) and common used configs (i.e., debug configs, logging configs, etc,) are grouped together and placed under `otaclient.configs` package.

The motivation for this change is that configs itself should be the top-level dependency(a.k.a, only other internal modules depends on it, it itself does't depend on any other internal modules). Previously `otaclient.app.configs` holds all configs, including app specific one and common shared one like logging settings, this is not ideal as `configs` module itself should not be a sub-module under `otaclient.app` as `otaclient.app` depends on it.
With moving configs to standalone `otaclient.configs`, the dependency relationship is fixed.

### pydantic as configs storage back-end

With `pydantic` as configs storing back-end, validation to otaclient configs are introduced and enforced. Other features from pydantic are also become available for Configs classes instances.

### User configurable configs can be configured via env var

With `pydantic` and `pydantic-settings`, some of the configs(configs in `app_cfg.ConfigurableConfigs`, `debug_cfg` and `logging_cfg`) that configure otaclient's runtime behavior are exposed to end-user and configurable via environmental variables.

To set `<config_name>` runtime configurable config, set the `OTA_<config_name>` environmental variable. 
For example, to configure enabling the debug mode, set `OTA_DEBUG_MODE=true` env.

For more details, please refer to `pydantic-settings` [documentation](https://docs.pydantic.dev/latest/concepts/pydantic_settings/#usage) and `app.configs` module.

### Rootfs specifying support and dynamic rooted path constants support

> [!Note]
> Actual container mode support is not yet implemented! This PR is the base requirement for container mode.

Now for otaclient running under container environment, the actual host rootfs mount point can be specified by `OTA_HOST_ROOTFS`, allowing otaclient to properly calling commands and executables from the host. 

Also, path constants are now dynamically rooted, configured automatically according to the value of `OTA_HOST_ROOTFS` during configs initializing.

## Check list

- [x] documentation for the new config infra.
- [x] local test is passed. 
- [x] VM(with grub) test is passed. 

## Major Changes

1. `otaclient.configs`: implement new configs infrastructure using pydantic, move `app.configs` into this package, split some configs from old `app.configs` into standalone modules `debug_cfg`, `logging_cfg` and `ota_service_cfg`,
2. `app.configs`: become a virtual module that imports needed configs from `otaclient.configs`,
3. `boot_control.configs`: adjust to use new `app.configs`, also refactor to use pydantic,
4. `boot_control.configs`: refine configs definition for each boot controller types,
5. `boot_control`: adjust `cboot`, `grub`, `rpi_boot` implementation to use new config infrastructure,
6. `app.main`: add a check to `ACTIVE_ROOTFS` runtime configurable config during starting,
7. `app.configs`: split `DEBUG_MODE` into one general switch and multiple sub debug feature flag, currently only `DEBUG_ENABLE_TRACEBACK_IN_STATUS_API` is defined.
8. new `otaclient.configs` package is defined to hold `otaclient` package scope configs.

## Behavior changes

Does this PR introduce behavior change(s)?

- [x] Yes, internal behaivor (will not impact user experience).
- [x] Yes, external behaivor (will impact user experience).

### Behavior with this PR

1. some of the otaclient configs(exposed runtime configurable configs) now are configurable via environmental variables,
9. validation to config values are introduced, for runtime configurable configs, invalid value will prevent otaclient from starting,
10.  actual host rootfs mount point can be specified now, all paths related constants are automatically dynamically configured accordingly.  

## Breaking change

Does this PR introduce breaking change?
- [x] No.

## Related links & tickets

[RT4-7156](https://tier4.atlassian.net/browse/RT4-7156)

[RT4-7156]: https://tier4.atlassian.net/browse/RT4-7156?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: otaclient/configs/app_cfg.py
Comment (Bodong-Yang) [inline otaclient/configs/app_cfg.py:1]: Moved from `app.configs` to `otaclient.configs.app_cfg`.

**Review by obi-t4 — APPROVED** [review]
Seems to be OK.

File: otaclient/_utils/__init__.py [hunk 1]
```diff
@@ -11,34 +11,29 @@
...
+import os.path
+from functools import cached_property
+from pydantic import computed_field
+from typing import Any, Callable, TypeVar
 
+T = TypeVar("T")
 
-RT = TypeVar("RT")
+_CONTAINER_INDICATOR_FILES = [
+    "/.dockerenv",
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/_utils/__init__.py:25]: Ah the OTA image is built under docker so this file is saved! Emmm actually this check is only for warning purpose(otaclient will issue warning if it detects itself is running in container but required ENV is not specified).

I am thinking of just removing this warning, and tell the user to ensure correct running mode and specify required ENV by themselves.

File: otaclient/app/boot_control/_grub.py [hunk 2]
```diff
@@ -744,12 +742,19 @@ class GrubController(BootControllerProtocol):
     def __init__(self) -> None:
         try:
             self._boot_control = _GrubControl()
+
+            # ------ prepare mount space ------ #
+            otaclient_ms = Path(cfg.OTACLIENT_MOUNT_SPACE_DPATH)
+            otaclient_ms.mkdir(exist_ok=True, parents=True)
+            otaclient_ms.chmod(0o700)
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/app/boot_control/_grub.py:749]: By default it should be 755(umask 022). Here I just want to make sure only `root` permission user can access this folder.

**Pull request merged**